File: ChessLab/OpeningGraph/OpeningCourseValidator.swift
import ChessKit
import Foundation

/// Vérifie l'INTÉGRITÉ d'un cours en graphe. Utilisé par les tests (aucune
/// arête orpheline, aucun FEN invalide, transpositions cohérentes) ET,
/// plus tard, par le générateur hors app comme garde-fou avant publication
/// d'un fichier.
///
/// PUR (cours → liste de problèmes), sans SwiftData ni UI. La validation des
/// arêtes rejoue chaque coup avec ChessKit et compare la clé normalisée
/// obtenue à `toFEN` : c'est le test le plus fort, il valide à la fois la
/// légalité du coup, la cohérence du graphe ET la normalisation FEN de bout
/// en bout.
enum OpeningCourseValidator {

    struct Issue: Equatable, CustomStringConvertible {
        enum Kind: String {
            case rootMissing            // rootFEN absent du graphe
            case keyMismatch            // clé du dictionnaire ≠ node.fen
            case invalidFEN             // FEN illisible par ChessKit
            case nonNormalizedFEN       // la clé n'est pas sa propre forme normalisée
            case orphanEdge             // arête pointant vers un nœud inexistant
            case illegalMove            // SAN/UCI non jouable depuis la position source
            case edgeTargetMismatch     // le coup mène à une position ≠ toFEN
            case chapterPositionMissing // un chapitre référence une position absente
        }
        let kind: Kind
        let detail: String
        var description: String { "[\(kind.rawValue)] \(detail)" }
    }

    /// Retourne tous les problèmes détectés (vide = graphe sain).
    static func validate(_ course: OpeningCourse) -> [Issue] {
        var issues: [Issue] = []

        if course.positions[course.rootFEN] == nil {
            issues.append(Issue(kind: .rootMissing, detail: "rootFEN absent : \(course.rootFEN)"))
        }

        for (key, node) in course.positions {
            if key != node.fen {
                issues.append(Issue(kind: .keyMismatch, detail: "clé \(key) ≠ node.fen \(node.fen)"))
            }
            guard OpeningFENKey.position(from: node.fen) != nil else {
                issues.append(Issue(kind: .invalidFEN, detail: node.fen))
                continue
            }
            if OpeningFENKey.normalize(node.fen) != node.fen {
                issues.append(Issue(kind: .nonNormalizedFEN, detail: node.fen))
            }
            for edge in node.moves {
                validateEdge(edge, from: node, in: course, into: &issues)
            }
        }

        for chapter in course.chapters ?? [] {
            for fen in chapter.positionFENs where course.positions[fen] == nil {
                issues.append(Issue(
                    kind: .chapterPositionMissing,
                    detail: "chapitre « \(chapter.title) » → position absente : \(fen)"
                ))
            }
        }

        return issues
    }

    private static func validateEdge(
        _ edge: MoveEdge, from node: PositionNode, in course: OpeningCourse, into issues: inout [Issue]
    ) {
        if course.positions[edge.toFEN] == nil {
            issues.append(Issue(kind: .orphanEdge, detail: "\(node.fen) --\(edge.san)--> \(edge.toFEN) (cible absente)"))
        }
        guard let resultingKey = resultingKey(afterUCI: edge.uci, from: node.fen) else {
            issues.append(Issue(kind: .illegalMove, detail: "\(edge.san)/\(edge.uci) illégal depuis \(node.fen)"))
            return
        }
        if resultingKey != edge.toFEN {
            issues.append(Issue(
                kind: .edgeTargetMismatch,
                detail: "\(node.fen) --\(edge.uci)--> obtenu \(resultingKey), attendu \(edge.toFEN)"
            ))
        }
    }

    /// Rejoue un coup UCI depuis une FEN et retourne la clé normalisée de la
    /// position obtenue, ou `nil` si le coup est illégal/illisible. Gère la
    /// promotion (5ᵉ caractère UCI, défaut dame), même mécanique que
    /// ``OpeningLineTrainingViewModel``.
    static func resultingKey(afterUCI uci: String, from fen: String) -> String? {
        guard uci.count >= 4, let position = OpeningFENKey.position(from: fen) else { return nil }
        let start = Square(String(uci.prefix(2)))
        let end = Square(String(uci.dropFirst(2).prefix(2)))

        var board = Board(position: position)
        guard board.canMove(pieceAt: start, to: end), let move = board.move(pieceAt: start, to: end) else {
            return nil
        }
        if case .promotion = board.state {
            // UCI promeut en minuscule (« e7e8q ») ; `Piece.Kind` a des
            // rawValues MAJUSCULES (N/B/R/Q) — d'où le `.uppercased()`.
            let kind: Piece.Kind = uci.count == 5
                ? (Piece.Kind(rawValue: String(uci.suffix(1)).uppercased()) ?? .queen)
                : .queen
            board.completePromotion(of: move, to: kind)
        }
        return OpeningFENKey.key(for: board.position)
    }
}

File: android/app/src/main/java/com/chesslab/courses/OpeningCourseValidator.kt
package com.chesslab.courses

import chesskit.Board
import chesskit.Piece
import chesskit.Square

/**
 * Vérifie l'INTÉGRITÉ d'un cours en graphe. Pendant d'`OpeningCourseValidator`.
 *
 * PUR (cours → liste de problèmes). La validation des arêtes REJOUE chaque coup
 * et compare la clé canonique obtenue à `toFEN` : c'est le test le plus fort,
 * il valide à la fois la légalité du coup, la cohérence du graphe et la
 * normalisation des clés de bout en bout. Un graphe incohérent ne doit pas
 * pouvoir entrer par la porte utilisateur alors qu'il est interdit par la
 * porte des assets.
 */
object OpeningCourseValidator {

    enum class Kind {
        /** `rootFEN` absent du graphe. */
        rootMissing,
        /** Une clé illisible. */
        invalidFEN,
        /** Une clé qui n'est pas sa propre forme canonique. */
        nonNormalizedFEN,
        /** Une arête vers un nœud inexistant. */
        orphanEdge,
        /** Un coup injouable depuis sa position. */
        illegalMove,
        /** Un coup qui ne mène pas où l'arête le dit. */
        edgeTargetMismatch,
        /** Un chapitre qui cite une position absente. */
        chapterPositionMissing,
    }

    data class Issue(val kind: Kind, val detail: String) {
        override fun toString() = "[$kind] $detail"
    }

    /** Tous les problèmes détectés — vide = graphe sain. */
    fun validate(course: Course): List<Issue> {
        val issues = ArrayList<Issue>()
        val root = CourseRepository.fenKey(course.rootFEN)
        if (root !in course.positions) issues += Issue(Kind.rootMissing, "rootFEN absent : $root")

        for ((key, moves) in course.positions) {
            val position = CourseRepository.position(key)
            if (position == null) { issues += Issue(Kind.invalidFEN, key); continue }
            if (CourseRepository.key(position) != key) issues += Issue(Kind.nonNormalizedFEN, key)
            for (edge in moves) {
                val to = CourseRepository.fenKey(edge.toFEN)
                if (to !in course.positions) issues += Issue(Kind.orphanEdge, "$key --${edge.san}--> $to (cible absente)")
                val landed = resultingKey(edge.uci, key)
                if (landed == null) { issues += Issue(Kind.illegalMove, "${edge.san}/${edge.uci} illégal depuis $key"); continue }
                if (landed != to) issues += Issue(Kind.edgeTargetMismatch, "$key --${edge.uci}--> obtenu $landed, attendu $to")
            }
        }

        for (chapter in course.chapters) {
            for (fen in chapter.positionFENs) {
                if (CourseRepository.fenKey(fen) !in course.positions) {
                    issues += Issue(Kind.chapterPositionMissing, "chapitre « ${chapter.title} » → position absente : $fen")
                }
            }
        }
        return issues
    }

    /**
     * Rejoue un coup UCI depuis une clé FEN et rend la clé CANONIQUE de la
     * position obtenue — `null` si le coup est illégal ou illisible. La
     * promotion suit le 5ᵉ caractère UCI, dame par défaut.
     */
    fun resultingKey(uci: String, fen: String): String? {
        if (uci.length < 4) return null
        val position = CourseRepository.position(fen) ?: return null
        val board = Board(position)
        val move = board.move(Square(uci.substring(0, 2)), Square(uci.substring(2, 4))) ?: return null
        if (board.state is Board.State.Promotion) {
            val kind = when (uci.getOrNull(4)) {
                'r' -> Piece.Kind.rook; 'b' -> Piece.Kind.bishop; 'n' -> Piece.Kind.knight; else -> Piece.Kind.queen
            }
            board.completePromotion(move, kind)
        }
        return CourseRepository.key(board.position)
    }
}
